### PR TITLE
Add client rack-awareness support to Mirror Maker 2

### PR DESCRIPTION
**Pull request by fvaleri**
### Type of change

_Select the type of your PR_

- Enhancement / new feature

### Description

This PR adds new property `client.rack` for Mirror Maker 2 cluster, which allows to use fetch from closest replica feature.

It implements https://github.com/strimzi/strimzi-kafka-operator/issues/6611.

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md

### CODE REVIEW

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaConnect.java [hunk 1]
```diff
@@ -185,4 +185,10 @@ public String toString() {
     public void setAdditionalProperty(String name, Object value) {
         this.additionalProperties.put(name, value);
     }
+
+    @Override
+    @Description("Rack awareness configuration.")
+    public Rack getRack() {
+        return spec.getRack();
+    }
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaConnect.java:193]: This seems wrong. This is a property of the spec not of the CRD.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java [hunk 2]
```diff
@@ -1800,4 +1802,31 @@ protected VolumeMount createTempDirVolumeMount() {
...
+                .build();
+
+        RoleRef roleRef = new RoleRefBuilder()
+                .withName("strimzi-kafka-client")
+                .withApiGroup("rbac.authorization.k8s.io")
+                .withKind("ClusterRole")
+                .build();
+
+        return getClusterRoleBinding(name, subject, roleRef);
+    }
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java:1831]: This abstraction is broken. I don't think it should be there since not everything inheriting from here has anything to do with "client" cluster role bindings.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperatorConfig.java [hunk 3]
```diff
@@ -89,6 +89,7 @@
     public static final boolean DEFAULT_NETWORK_POLICY_GENERATION = true;
     public static final boolean DEFAULT_CREATE_CLUSTER_ROLES = false;
     public static final boolean DEFAULT_POD_SET_RECONCILIATION_ONLY = false;
+    public static final String DEFAULT_STRIMZI_OPERATOR_IMAGE = "quay.io/strimzi/operator:latest";
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperatorConfig.java:92]: How is this and the uses of this variable related to this PR?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java [hunk 4]
```diff
@@ -833,32 +829,6 @@ public Tracing getTracing() {
         return tracing;
     }
 
-    /**
-     * Creates the ClusterRoleBinding which is used to bind the Kafka Connect SA to the ClusterRole
-     * which permissions the Kafka init container to access K8S nodes (necessary for rack-awareness).
-     *
-     * @return The cluster role binding.
-     */
-    public ClusterRoleBinding generateClusterRoleBinding() {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java:842]: Since KMM2 inherits from KafkaConnect, why does this need to be removed / moved?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java [hunk 5]
```diff
@@ -43,27 +18,50 @@
...
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java:64]: Can we keep the ordering as it was?

File: packaging/helm-charts/helm3/strimzi-kafka-operator/crds/041-Crd-kafkaconnect.yaml [hunk 6]
```diff
@@ -1910,3 +1910,13 @@ spec:
                   type: integer
                   description: The current number of pods being used to provide this resource.
               description: The status of the Kafka Connect cluster.
+            rack:
```
Comment (scholzj) [inline packaging/helm-charts/helm3/strimzi-kafka-operator/crds/041-Crd-kafkaconnect.yaml:1913]: This seems wrong. The rack section already is here. This file should not change.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperatorConfig.java (hunk 3 above)
Comment (fvaleri) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperatorConfig.java:92]: This is just to have a single place for default image URL.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java (hunk 4 above)
Comment (fvaleri) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java:842]: OK, thanks.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java (hunk 2 above)
Comment (fvaleri) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java:1831]: OK, thanks.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaConnect.java (hunk 1 above)
Comment (fvaleri) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaConnect.java:193]: OK, thanks.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java (hunk 5 above)
Comment (fvaleri) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java:64]: I'm reverting back to the original ordering. This is idea auto-import.

File: packaging/helm-charts/helm3/strimzi-kafka-operator/crds/041-Crd-kafkaconnect.yaml (hunk 6 above)
Comment (fvaleri) [reply packaging/helm-charts/helm3/strimzi-kafka-operator/crds/041-Crd-kafkaconnect.yaml:1913]: Right, too much refactoring.

**Comment by fvaleri** [comment]
@scholzj I fixed the breaking changes.
Converted to draft as I want to do some more tests.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperatorConfig.java (hunk 3 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperatorConfig.java:92]: Yeah, I understood it. But having a separate PR for it would make it much easier to review things

File: documentation/api/io.strimzi.api.kafka.model.Rack.adoc [hunk 7]
```diff
@@ -89,12 +92,23 @@ apiVersion: {KafkaConnectApiVersion}
...
+----
+apiVersion: {KafkaMirrorMaker2ApiVersion}
+kind: KafkaMirrorMaker2
+# ...
+spec:
+  # ...
+  rack:
+    topologyKey: topology.kubernetes.io/zone
+  # ...
+----
```
Comment (scholzj) [inline documentation/api/io.strimzi.api.kafka.model.Rack.adoc:114]: What is the actual impact on the different MM2 connectors? Is this something what is desired for all of them?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMaker2Cluster.java [hunk 8]
```diff
@@ -142,9 +146,16 @@ private static KafkaConnectSpec buildKafkaConnectSpec(KafkaMirrorMaker2Spec spec
...
                 .withTemplate(spec.getTemplate())
-                .withExternalConfiguration(spec.getExternalConfiguration())
-                .build();
+                .withExternalConfiguration(spec.getExternalConfiguration());
+
+        Rack rack = spec.getRack();
+        if (rack != null) {
+            kafkaConnectSpecBuilder
+                    .withRack(rack)
+                    .withImage(System.getenv().getOrDefault(STRIMZI_DEFAULT_KAFKA_INIT_IMAGE, DEFAULT_STRIMZI_OPERATOR_IMAGE));
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMaker2Cluster.java:155]: `withClientRackInitImage`?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMaker2Cluster.java [hunk 9]
```diff
@@ -142,9 +146,16 @@ private static KafkaConnectSpec buildKafkaConnectSpec(KafkaMirrorMaker2Spec spec
...
+        }
 
+        return kafkaConnectSpecBuilder.build();
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMaker2Cluster.java:158]: Is this needed? Why can't you just do `withRack(spec.getRack)` and the same for the image? It would be either `null` or have the right value

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java [hunk 10]
```diff
@@ -834,13 +835,14 @@ public Tracing getTracing() {
...
-     * which permissions the Kafka init container to access K8S nodes (necessary for rack-awareness).
+     * Creates a ClusterRoleBinding which is used to bind the SA to the strimzi-kafka-client ClusterRole.
+     *
+     * @param name Name of the cluster role binding.
      *
      * @return The cluster role binding.
      */
-    public ClusterRoleBinding generateClusterRoleBinding() {
-        if (rack == null) {
+    public ClusterRoleBinding generateKafkaClientClusterRoleBinding(String name) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java:844]: I think this is weird. I understand that you probably want to share the method but use two different names. But I'm not sure if it should be part of the public API. Maybe you can have protected method for the CRB with the name parameter and public ones in the model classes without the name which would use their own name to call the protected methods.

Basing the existence of the resource on the name seems also completely counterintuitive. It should continue use the `rack` field.

I also don't follow why this needed the renaming. Seems like unnecessary change. The old Javadoc seemed to be also more descriptive.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java [hunk 11]
```diff
@@ -103,14 +101,14 @@ public KafkaConnectAssemblyOperator(Vertx vertx, PlatformFeaturesAvailability pf
...
+        String clusterRoleBindingName = KafkaConnectResources.initContainerClusterRoleBindingName(kafkaConnect.getMetadata().getName(), namespace);
         boolean connectHasZeroReplicas = connect.getReplicas() == 0;
-
         final AtomicReference<String> image = new AtomicReference<>();
         final AtomicReference<String> desiredLogging = new AtomicReference<>();
+
+        LOGGER.debugCr(reconciliation, "Updating Kafka Connect cluster");
         connectServiceAccount(reconciliation, namespace, KafkaConnectResources.serviceAccountName(connect.getCluster()), connect)
-                .compose(i -> connectInitClusterRoleBinding(reconciliation, namespace, kafkaConnect.getMetadata().getName(), connect))
+                .compose(i -> clusterRoleBindingForRackAwareness(reconciliation, clusterRoleBindingName, kafkaConnect.getSpec(), connect))
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java:111]: Again, why renaming it and causing unnecessary changes?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java [hunk 12]
```diff
@@ -892,4 +893,26 @@ public Timer connectorsReconciliationsTimer(String namespace) {
...
+     * The init-container needs to be able to read the labels from the node it is running on to be able
+     * to determine the `client.rack` option.
+     *
+     * @param reconciliation The reconciliation
+     * @param name ClusterRoleBinding name
+     * @param spec Cluster spec
+     * @param cluster Cluster instance
+     * @return Future for tracking the asynchronous result of the ClusterRoleBinding reconciliation
+     */
+    protected Future<ReconcileResult<ClusterRoleBinding>> clusterRoleBindingForRackAwareness(Reconciliation reconciliation,
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java:907]: Why doesn't this follow the old name `connectInitClusterRoleBinding`?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperatorConfig.java (hunk 3 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperatorConfig.java:92]: While the change it self might make sense, I think I counted at least 8 files changed only by this. It complicates the review and makes it harder to track changes in the future.

**Review by scholzj — COMMENTED** [review]
I left some more comments. But the APi changes look good now. I think we should think about a ST for this as well -> at least a simple one that shows that the deployment with the rack proceeds and works.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperatorConfig.java (hunk 3 above)
Comment (fvaleri) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/ClusterOperatorConfig.java:92]: OK.

**Comment by fvaleri** [comment]
I added the missing peace to make it work, and it's now ready for another round of review. As you can see from the following log, MirrorSourceConnector is always fetching from its own zone replica:

```shell
2022-04-13 10:57:13,815 DEBUG [my-cluster-src->my-cluster-tgt.MirrorSourceConnector|task-0] [Consumer clientId=consumer-null-4, groupId=null] Added READ_UNCOMMITTED fetch request for partition my-topic-2 at position FetchPosition{offset=0, offsetEpoch=Optional.empty, currentLeader=LeaderAndEpoch{leader=Optional[my-cluster-src-kafka-0.my-cluster-src-kafka-brokers.kafka-src.svc:9093 (id: 0 rack: us-east-1a)], epoch=0}} to node my-cluster-src-kafka-2.my-cluster-src-kafka-brokers.kafka-src.svc:9093 (id: 2 rack: us-east-1c) (org.apache.kafka.clients.consumer.internals.Fetcher) [task-thread-my-cluster-src->my-cluster-tgt.MirrorSourceConnector-0]
2022-04-13 10:57:13,815 DEBUG [my-cluster-src->my-cluster-tgt.MirrorSourceConnector|task-0] [Consumer clientId=consumer-null-4, groupId=null] Added READ_UNCOMMITTED fetch request for partition my-topic-0 at position FetchPosition{offset=100, offsetEpoch=Optional[0], currentLeader=LeaderAndEpoch{leader=Optional[my-cluster-src-kafka-1.my-cluster-src-kafka-brokers.kafka-src.svc:9093 (id: 1 rack: us-east-1b)], epoch=0}} to node my-cluster-src-kafka-2.my-cluster-src-kafka-brokers.kafka-src.svc:9093 (id: 2 rack: us-east-1c) (org.apache.kafka.clients.consumer.internals.Fetcher) [task-thread-my-cluster-src->my-cluster-tgt.MirrorSourceConnector-0]
```

If it looks good, then I can implement my test procedure as a proper system test.

File: documentation/api/io.strimzi.api.kafka.model.Rack.adoc (hunk 7 above)
Comment (fvaleri) [reply documentation/api/io.strimzi.api.kafka.model.Rack.adoc:114]: This is mainly for MirrorSourceConnector which creates a consumer inside its task. The other two don't, so it's simply ignored.

**Review by scholzj — COMMENTED** [review]
Sone nits, but looks good in general. Have you thought about an ST for this? Or should we leave it to the QE folks?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java [hunk 13]
```diff
@@ -892,4 +893,22 @@ public Timer connectorsReconciliationsTimer(String namespace) {
...
+     * @param cluster Cluster instance
+     * @return Future for tracking the asynchronous result of the ClusterRoleBinding reconciliation
+     */
+    protected Future<ReconcileResult<ClusterRoleBinding>> connectInitClusterRoleBinding(Reconciliation reconciliation,
+                                                                                        String name,
+                                                                                        AbstractKafkaConnectSpec spec,
+                                                                                        KafkaConnectCluster cluster) {
+        ClusterRoleBinding crb = spec.getRack() != null ? cluster.generateClusterRoleBinding(spec.getRack(), name) : null;
+        return withIgnoreRbacError(reconciliation, clusterRoleBindingOperations.reconcile(reconciliation, name, crb), crb);
+    }
```
[the remark this answers is not in the document]
Comment (fvaleri) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java:913]: Agree, I was more focused on making it work. Will improve that.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java [hunk 14]
```diff
@@ -144,9 +145,11 @@ protected Future<KafkaMirrorMaker2Status> createOrUpdate(Reconciliation reconcil
         final AtomicReference<String> desiredLogging = new AtomicReference<>();
 
         boolean mirrorMaker2HasZeroReplicas = mirrorMaker2Cluster.getReplicas() == 0;
+        String clusterRoleBindingName = KafkaMirrorMaker2Resources.initContainerClusterRoleBindingName(kafkaMirrorMaker2.getMetadata().getName(), namespace);
```
[the remark this answers is not in the document]
Comment (fvaleri) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java:148]: I would keep this as it is for readability. The JVM will inline at runtime.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java [hunk 15]
```diff
@@ -837,10 +837,13 @@ public Tracing getTracing() {
...
      *
+     * @param rack Rack configuration.
+     * @param name ClusterRoleBinding name.
+     *
      * @return The cluster role binding.
      */
-    public ClusterRoleBinding generateClusterRoleBinding() {
-        if (rack == null) {
+    public ClusterRoleBinding generateClusterRoleBinding(Rack rack, String name) {
+        if (rack == null || name == null || name.trim().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (fvaleri) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java:846]: I think passing the rack explicitly is clearer. I'll remove the name that is not needed.

[the remark this answers is not in the document]
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java:846]: Why is it clearer when it is part of the model object to which this method belongs anyway? That is why it was designed around the class hierarchy to use these things internally. Without it it can be all rewritten as static methods and everything can be passwed around in parameters.

[the remark this answers is not in the document]
Comment (fvaleri) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java:846]: Right, I see it now. Let mi fix that.

**Comment by fvaleri** [comment]
Done requested refactoring and cleanup. 

I already have a test procedure that I'll implement as a system test in the following days.

**Review by scholzj — APPROVED** [review]

File: api/src/main/java/io/strimzi/api/kafka/model/AbstractKafkaConnectSpec.java [hunk 16]
```diff
@@ -178,4 +181,24 @@ public ExternalConfiguration getExternalConfiguration() {
...
+    @JsonInclude(value = JsonInclude.Include.NON_NULL)
+    public String getClientRackInitImage() {
+        return clientRackInitImage;
+    }
+
+    public void setClientRackInitImage(String brokerRackInitImage) {
+        this.clientRackInitImage = brokerRackInitImage;
+    }
+
+    @Description("Configuration of the node label which will be used as the client.rack consumer configuration.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/AbstractKafkaConnectSpec.java:195]:
```suggestion
    @Description("Configuration of the node label which will be used as the `client.rack` consumer configuration.")
```

File: documentation/api/io.strimzi.api.kafka.model.Rack.adoc [hunk 17]
```diff
@@ -79,7 +79,10 @@ If the rack ID is not specified, or if it cannot find a replica with the same ra
 .Example showing client consuming from replicas in the same availability zone
 image::rack-config-availability-zones.png[consuming from replicas in the same availability zone]
 
-Consuming messages from the closest replicas can be used also in Kafka Connect for sink connectors which are consuming messages.
+Consuming messages from the closest replicas can be used also in Kafka Connect and Mirror Maker 2 for connectors which are consuming messages.
+Enabling rack awareness in the `KafkaConnect` and `KafkaMirrorMaker2` custom resource will not set any affinity rules, but you can also configure `affinity` or `topologySpreadConstraints`.
+For more information see xref:assembly-scheduling-str[].
```
Comment (PaulRMellor) [inline documentation/api/io.strimzi.api.kafka.model.Rack.adoc:84]:
```suggestion
You can also configure Kafka Connect and MirrorMaker 2.0 so that connectors consume messages from the closest replicas.  
You enable rack awareness in the `KafkaConnect` and `KafkaMirrorMaker2` custom resources. 
The configuration does does not set affinity rules, but you can also configure `affinity` or `topologySpreadConstraints`.
For more information see xref:assembly-scheduling-str[].
```

File: documentation/api/io.strimzi.api.kafka.model.Rack.adoc [hunk 18]
```diff
@@ -89,12 +92,23 @@ apiVersion: {KafkaConnectApiVersion}
...
-    # ...
+  # ...
+  rack:
+    topologyKey: topology.kubernetes.io/zone
+  # ...
 ----
 
-Enabling rack awareness in the `KafkaConnect` custom resource will not set any affinity rules, but you can also configure `affinity` or `topologySpreadConstraints`.
-For more information see xref:assembly-scheduling-str[].
+When deploying Mirror Maker 2 using Strimzi, you can use the `rack` section in the `KafkaMirrorMaker2` custom resource to automatically configure the `client.rack` option.
```
Comment (PaulRMellor) [inline documentation/api/io.strimzi.api.kafka.model.Rack.adoc:101]:
```suggestion
When deploying MirrorMaker 2.0 using Strimzi, you can use the `rack` section in the `KafkaMirrorMaker2` custom resource to automatically configure the `client.rack` option.
```

**Review by PaulRMellor — APPROVED** [review]
Looks good. Left a couple of suggestions for the doc.

**Comment by fvaleri** [comment]
Thanks @PaulRMellor.

**Comment by scholzj** [comment]
@fvaleri Looks like you need to update the derived resources:

```
ERROR: Uncommitted changes in derived resources:
M	documentation/modules/appendix_crds.adoc
M	packaging/helm-charts/helm3/strimzi-kafka-operator/crds/041-Crd-kafkaconnect.yaml
M	packaging/helm-charts/helm3/strimzi-kafka-operator/crds/048-Crd-kafkamirrormaker2.yaml
M	packaging/install/cluster-operator/041-Crd-kafkaconnect.yaml
M	packaging/install/cluster-operator/048-Crd-kafkamirrormaker2.yaml
```

File: api/src/main/java/io/strimzi/api/kafka/model/AbstractKafkaConnectSpec.java [hunk 19]
```diff
@@ -43,6 +44,8 @@ public abstract class AbstractKafkaConnectSpec extends Spec implements HasConfig
     private Tracing tracing;
     private KafkaConnectTemplate template;
     private ExternalConfiguration externalConfiguration;
+    private String clientRackInitImage;
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/AbstractKafkaConnectSpec.java:47]: I was wondering if we want to be so explicit about the name of this init image used just for client rack.
I mean, now it's just client rack but who knows the future and if that could be used for some other pre-processing.
Can't we think about a more generic name? Just `initImage` or something like that?

Comment (scholzj) [reply api/src/main/java/io/strimzi/api/kafka/model/AbstractKafkaConnectSpec.java:47]: You cannot change this. It is part of the Connect API already.

Comment (fvaleri) [reply api/src/main/java/io/strimzi/api/kafka/model/AbstractKafkaConnectSpec.java:47]: That's true, and it is the reason I didn't touch it. We could create a better name deprecating the old one, but I would do it only when we will actually need to do multiple things during the init phase.

Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/AbstractKafkaConnectSpec.java:47]: Damn! How was it used and why named "clientRack" if we are adding the rack-awareness now?

**Review by ppatierno — APPROVED** [review]

File: api/src/main/java/io/strimzi/api/kafka/model/AbstractKafkaConnectSpec.java (hunk 19 above)
Comment (fvaleri) [reply api/src/main/java/io/strimzi/api/kafka/model/AbstractKafkaConnectSpec.java:47]: MM2 rack awareness feature is built on top of KafkaConnect's one, that's why.

Comment (scholzj) [reply api/src/main/java/io/strimzi/api/kafka/model/AbstractKafkaConnectSpec.java:47]: Connect supports it for a long time. This PR only adds the support to MM2. In theory, you can have the configuration completely separate and use different options for each. But given that we do not have any plans for the init container to do anything else (I mean not just no plans to add something this month, but I'm not aware of any feature which would do it) I would stick with what Federico did.

**Comment by scholzj** [comment]
Thanks for the PR!
